File: protocol/src/main/resources/csharp/ProtocolRegistrationTemplate.cs
public class ${protocol_name}Registration : IProtocolRegistration
{
    public short ProtocolId()
    {
        return ${protocol_id};
    }

    public void Write(ByteBuffer buffer, object packet)
    {
        if (packet == null)
        {
            buffer.WriteInt(0);
            return;
        }
        ${protocol_name} message = (${protocol_name}) packet;
        ${protocol_write_serialization}
    }

    public object Read(ByteBuffer buffer)
    {
        int length = buffer.ReadInt();
        if (length == 0)
        {
            return null;
        }
        int beforeReadIndex = buffer.ReadOffset();
        ${protocol_name} packet = new ${protocol_name}();
        ${protocol_read_deserialization}
        if (length > 0)
        {
            buffer.SetReadOffset(beforeReadIndex + length);
        }
        return packet;
    }
}

File: protocol/src/main/resources/java/ProtocolRegistrationTemplate.java
class ${protocol_name}Registration implements IProtocolRegistration {
    @Override
    public short protocolId() {
        return ${protocol_id};
    }

    @Override
    public void write(ByteBuffer buffer, Object packet) {
        if (packet == null) {
            buffer.writeInt(0);
            return;
        }
        ${protocol_name} message = (${protocol_name}) packet;
        ${protocol_write_serialization}
    }

    @Override
    public Object read(ByteBuffer buffer) {
        var length = buffer.readInt();
        if (length == 0) {
            return null;
        }
        var beforeReadIndex = buffer.getReadOffset();
        var packet = new ${protocol_name}();
        ${protocol_read_deserialization}
        if (length > 0) {
            buffer.setReadOffset(beforeReadIndex + length);
        }
        return packet;
    }
};